File: dotnet/DesignPatterns/CreationalDesignPatterns/BuilderDesignPattern/BuilderDesignPattern.Example2/Pizza.cs
namespace BuilderDesignPattern.Example2;

public class Pizza
{
	public string Size { get; private set; }
	public string Dough { get; private set; }
	public string Sauce { get; private set; }
	public bool HasCheese { get; private set; }
	public bool HasPepperoni { get; private set; }
	public bool HasMushrooms { get; private set; }
	public bool HasOlives { get; private set; }

	private Pizza() { }

	public override string ToString()
	{
		return $"Pizza: Size={Size}, Dough={Dough}, Sauce={Sauce}, Cheese={HasCheese}, " +
			   $"Pepperoni={HasPepperoni}, Mushrooms={HasMushrooms}, Olives={HasOlives}";
	}

	public class PizzaBuilder
	{
		private readonly Pizza _pizza;

		public PizzaBuilder()
		{
			_pizza = new Pizza();
		}

		public PizzaBuilder SetSize(string size)
		{
			_pizza.Size = size;
			return this;
		}

		public PizzaBuilder SetDough(string dough)
		{
			_pizza.Dough = dough;
			return this;
		}

		public PizzaBuilder SetSauce(string sauce)
		{
			_pizza.Sauce = sauce;
			return this;
		}

		public PizzaBuilder AddCheese()
		{
			_pizza.HasCheese = true;
			return this;
		}

		public PizzaBuilder AddPepperoni()
		{
			_pizza.HasPepperoni = true;
			return this;
		}

		public PizzaBuilder AddMushrooms()
		{
			_pizza.HasMushrooms = true;
			return this;
		}

		public PizzaBuilder AddOlives()
		{
			_pizza.HasOlives = true;
			return this;
		}

		public Pizza Build()
		{
			if (string.IsNullOrWhiteSpace(_pizza.Size) || string.IsNullOrWhiteSpace(_pizza.Dough))
			{
				throw new InvalidOperationException("Pizza must have both Size and Dough set.");
			}

			return _pizza;
		}
	}
}

File: nodejs/DesignPatterns/CreationalDesignPatterns/BuilderDesignPattern/BuilderDesignPattern.Example2/src/Pizza.ts
export class Pizza {
  size: string = '';
  dough: string = '';
  sauce: string = '';
  hasCheese: boolean = false;
  hasPepperoni: boolean = false;
  hasMushrooms: boolean = false;
  hasOlives: boolean = false;

  private constructor() {}

  toString(): string {
    return (
      `Pizza: Size=${this.size}, Dough=${this.dough}, Sauce=${this.sauce}, ` +
      `Cheese=${this.hasCheese}, Pepperoni=${this.hasPepperoni}, ` +
      `Mushrooms=${this.hasMushrooms}, Olives=${this.hasOlives}`
    );
  }

  static Builder = class PizzaBuilder {
    private readonly _pizza: Pizza = new Pizza();

    setSize(size: string): this {
      this._pizza.size = size;
      return this;
    }

    setDough(dough: string): this {
      this._pizza.dough = dough;
      return this;
    }

    setSauce(sauce: string): this {
      this._pizza.sauce = sauce;
      return this;
    }

    addCheese(): this {
      this._pizza.hasCheese = true;
      return this;
    }

    addPepperoni(): this {
      this._pizza.hasPepperoni = true;
      return this;
    }

    addMushrooms(): this {
      this._pizza.hasMushrooms = true;
      return this;
    }

    addOlives(): this {
      this._pizza.hasOlives = true;
      return this;
    }

    build(): Pizza {
      if (!this._pizza.size || !this._pizza.dough) {
        throw new Error('Pizza must have both Size and Dough set.');
      }
      return this._pizza;
    }
  };
}
